src/twisted/internet/cfreactor.py

```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
         self._scheduleSimulate()
         return result
 
+    def startRunning(self, installSignalHandlers: bool = True) -> None:
+        """
+        Start running the reactor, then kick off the timer that advances
+        Twisted's clock to keep pace with CFRunLoop's.
+        """
+        super().startRunning(installSignalHandlers)
+        # We must force the simulator to run because in addition to scheduling
+        # timed calls, super().startRunning also fires startup events which may
+        # crash the reactor.
+        self._scheduleSimulate(force=True)
+
     _inCFLoop = False
 
-    def mainLoop(self):
+    def mainLoop(self) -> None:
         """
         Run the runner (C{CFRunLoopRun} or something that calls it), which runs
         the run loop until C{crash()} is called.
         """
-        self._inCFLoop = True
+        # we were crashed during startup.
+        if not self._started:
+
+            def docrash() -> None:
+                self.crash()
+
+            # we're running again, just for a moment, while we wait for some
+            # I/O to take place.  We need this to be True for callLater to not
```

Comment:
The "I/O" that we wait for is just the CF Timer to fire and call our crash function right?

Reply:
What I was trying to get across here is that we need to call into the actual loop, not just run the logic of the timer immediately. To allow some chance for already-ready file descriptors to be invoked. So the "I/O" is explicitly in contrast to this logic.

I'll try to make this clearer